File: tools/update_heads.py
import os
import re

ROOT = r'c:\xampp\htdocs\Library-Management-System'
SKIP = {'index.php', 'signup.php', 'admin/a.php', 'admin/b.php', 'tools/update_heads.py'}

def update_file(path, is_admin):
    with open(path, encoding='utf-8', errors='ignore') as f:
        content = f.read()

    if 'includes/head.php' in content:
        return False

    m = re.search(r'<head>\s*(.*?)\s*</head>', content, re.DOTALL | re.IGNORECASE)
    if not m:
        return False

    head_inner = m.group(1)
    title_m = re.search(r'<title>([^<]*)</title>', head_inner, re.I)
    page_title = title_m.group(1).strip() if title_m else 'NEW HARGEISA LIBRARY'
    use_dt = 'datatables.css' in head_inner

    include_path = 'includes/head.php' if not is_admin else 'includes/head.php'
    php_setup = "<?php\n$page_title = " + repr(page_title) + ";\n"
    if use_dt:
        php_setup += "$use_datatables = true;\n"
    php_setup += "?>"

    new_head = "<head>\n" + php_setup + "\n<?php include('" + include_path + "'); ?>\n</head>"
    new_content = content[:m.start()] + new_head + content[m.end():]

    # Remove inline style blocks in head area only (browse-books has style after head - handle separately)
    if new_content != content:
        with open(path, 'w', encoding='utf-8') as f:
            f.write(new_content)
        return True
    return False

for dirpath, _, files in os.walk(ROOT):
    for fn in files:
        if not fn.endswith('.php'):
            continue
        rel = os.path.relpath(os.path.join(dirpath, fn), ROOT).replace('\\', '/')
        if rel in SKIP or rel.startswith('tools/'):
            continue
        if 'config.php' in rel or 'header.php' in rel or 'head.php' in rel:
            continue
        if 'get_' in fn or fn in ('captcha.php', 'check_availability.php', 'temp.php', 'bgwork.php', 'logout.php', 'adminlogin.php'):
            continue
        is_admin = rel.startswith('admin/')
        if update_file(os.path.join(dirpath, fn), is_admin):
            print('updated', rel)

File: php/tools/update_heads.php
<?php
// PHP CLI utility to recursively scan and update legacy <head> blocks to unified PHP include statements
$root = dirname(__DIR__, 2);
$skip = ['index.php', 'signup.php', 'admin/a.php', 'admin/b.php', 'tools/update_heads.py', 'tools/update_heads.php', 'php/tools/update_heads.php'];

function updateFile($path, $isAdmin) {
    $content = file_get_contents($path);

    if (strpos($content, 'includes/head.php') !== false) {
        return false;
    }

    if (!preg_match('/<head>\s*(.*?)\s*<\/head>/is', $content, $matches)) {
        return false;
    }

    $headInner = $matches[1];
    $pageTitle = 'NEW HARGEISA LIBRARY';
    if (preg_match('/<title>([^<]*)<\/title>/i', $headInner, $titleMatch)) {
        $pageTitle = trim($titleMatch[1]);
    }
    
    $useDt = (strpos($headInner, 'datatables.css') !== false || strpos($headInner, 'dataTables.bootstrap.css') !== false);
    
    $includePath = $isAdmin ? 'includes/head.php' : 'includes/head.php';
    
    $phpSetup = "<?php\n\$page_title = " . var_export($pageTitle, true) . ";\n";
    if ($useDt) {
        $phpSetup .= "\$use_datatables = true;\n";
    }
    $phpSetup .= "?>";

    $newHead = "<head>\n" . $phpSetup . "\n<?php include('" . $includePath . "'); ?>\n</head>";
    $newContent = str_replace($matches[0], $newHead, $content);

    if ($newContent !== $content) {
        file_put_contents($path, $newContent);
        return true;
    }
    return false;
}

function scanDirRecursive($dir, $root, $skip) {
    $files = scandir($dir);
    foreach ($files as $file) {
        if ($file === '.' || $file === '..') {
            continue;
        }
        $path = $dir . '/' . $file;
        if (is_dir($path)) {
            scanDirRecursive($path, $root, $skip);
        } else {
            if (pathinfo($path, PATHINFO_EXTENSION) !== 'php') {
                continue;
            }
            $rel = str_replace('\\', '/', substr($path, strlen($root) + 1));
            if (in_array($rel, $skip) || strpos($rel, 'tools/') === 0 || strpos($rel, 'php/tools/') === 0) {
                continue;
            }
            if (strpos($rel, 'config.php') !== false || strpos($rel, 'header.php') !== false || strpos($rel, 'head.php') !== false) {
                continue;
            }
            if (strpos($file, 'get_') === 0 || in_array($file, ['captcha.php', 'check_availability.php', 'temp.php', 'bgwork.php', 'logout.php', 'adminlogin.php'])) {
                continue;
            }
            $isAdmin = (strpos($rel, 'php/admin/') === 0 || strpos($rel, 'admin/') === 0);
            if (updateFile($path, $isAdmin)) {
                echo "Updated: {$rel}\n";
            }
        }
    }
}

scanDirRecursive($root, $root, $skip);
echo "Scan and header standardizations completed successfully!\n";
?>
